# Your Title Here

**Name(s)**: (your name(s) here)

**Website Link**: (your website link)

## Code

In [6]:
import pandas as pd
import numpy as np
import os
import plotly.express as px
pd.options.plotting.backend = 'plotly'

#### Data Import

In [4]:
df = pd.read_excel('outage.xlsx').drop(['variables', 'OBS'], axis = 1)
df

YEAR  MONTH    U.S._STATE POSTAL.CODE NERC.REGION      CLIMATE.REGION  \
0     2011    7.0     Minnesota          MN         MRO  East North Central   
1     2014    5.0     Minnesota          MN         MRO  East North Central   
2     2010   10.0     Minnesota          MN         MRO  East North Central   
3     2012    6.0     Minnesota          MN         MRO  East North Central   
4     2015    7.0     Minnesota          MN         MRO  East North Central   
...    ...    ...           ...         ...         ...                 ...   
1529  2011   12.0  North Dakota          ND         MRO  West North Central   
1530  2006    NaN  North Dakota          ND         MRO  West North Central   
1531  2009    8.0  South Dakota          SD         RFC  West North Central   
1532  2009    8.0  South Dakota          SD         MRO  West North Central   
1533  2000    NaN        Alaska          AK        ASCC                 NaN   

      ANOMALY.LEVEL CLIMATE.CATEGORY OUTAGE.START.DATE OUTAGE.START.TIME  ...  \
0              -0.3           normal        2011-07-01          17:00:00  ...   
1              -0.1           normal        2014-05-11          18:38:00  ...   
2              -1.5             cold        2010-10-26          20:00:00  ...   
3              -0.1           normal        2012-06-19          04:30:00  ...   
4               1.2             warm        2015-07-18          02:00:00  ...   
...             ...              ...               ...               ...  ...   
1529           -0.9             cold        2011-12-06          08:00:00  ...   
1530            NaN              NaN               NaT               NaN  ...   
1531            0.5             warm        2009-08-29          22:54:00  ...   
1532            0.5             warm        2009-08-29          11:00:00  ...   
1533            NaN              NaN               NaT               NaN  ...   

     POPPCT_URBAN POPPCT_UC POPDEN_URBAN POPDEN_UC POPDEN_RURAL  \
0           73.27     15.28       2279.0    1700.5         18.2   
1           73.27     15.28       2279.0    1700.5         18.2   
2           73.27     15.28       2279.0    1700.5         18.2   
3           73.27     15.28       2279.0    1700.5         18.2   
4           73.27     15.28       2279.0    1700.5         18.2   
...           ...       ...          ...       ...          ...   
1529        59.90     19.90       2192.2    1868.2          3.9   
1530        59.90     19.90       2192.2    1868.2          3.9   
1531        56.65     26.73       2038.3    1905.4          4.7   
1532        56.65     26.73       2038.3    1905.4          4.7   
1533        66.02     21.56       1802.6    1276.0          0.4   

      AREAPCT_URBAN  AREAPCT_UC   PCT_LAND  PCT_WATER_TOT  PCT_WATER_INLAND  
0              2.14        0.60  91.592666       8.407334          5.478743  
1              2.14        0.60  91.592666       8.407334          5.478743  
2              2.14        0.60  91.592666       8.407334          5.478743  
3              2.14        0.60  91.592666       8.407334          5.478743  
4              2.14        0.60  91.592666       8.407334          5.478743  
...             ...         ...        ...            ...               ...  
1529           0.27        0.10  97.599649       2.401765          2.401765  
1530           0.27        0.10  97.599649       2.401765          2.401765  
1531           0.30        0.15  98.307744       1.692256          1.692256  
1532           0.30        0.15  98.307744       1.692256          1.692256  
1533           0.05        0.02  85.761154      14.238846          2.901182  

[1534 rows x 55 columns]

In [168]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1534 entries, 0 to 1533
Data columns (total 55 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   YEAR                     1534 non-null   int64         
 1   MONTH                    1525 non-null   float64       
 2   U.S._STATE               1534 non-null   object        
 3   POSTAL.CODE              1534 non-null   object        
 4   NERC.REGION              1534 non-null   object        
 5   CLIMATE.REGION           1528 non-null   object        
 6   ANOMALY.LEVEL            1525 non-null   float64       
 7   CLIMATE.CATEGORY         1525 non-null   object        
 8   OUTAGE.START.DATE        1525 non-null   datetime64[ns]
 9   OUTAGE.START.TIME        1525 non-null   object        
 10  OUTAGE.RESTORATION.DATE  1476 non-null   datetime64[ns]
 11  OUTAGE.RESTORATION.TIME  1476 non-null   object        
 12  CAUSE.CATEGORY           1534 non-

#### Introduction and Question Identification:

This dataset provides very detailed information about the power outages within US from January 2000 to July 2016, recording 1534 reports of power outage. The dataset mainly consists of 4 parts: general information, regional climate information, outage events information, and regional electricity consumption information. I've considered many questions to be asked about this dataset, including the places and time that major outage happens, the relationship between electricity price and probability of having an outage... But I found most of them to be hard to answer with a few hypothesis testing. Thus, the question we will disscuss here is how factors like season, climate, and cause of outage affect the seriousness of power outage, and we will answer this question by a few hypothesis testing in the end of the project. This anlaysis could be very important for electricity companies to identify potential problems and avoid serious power outage in their service. To answer this question, we will focus on these variables in our analysis:
<table>
        <thead>
            <tr>
                <th>Variable Name</th>
                <th>Description</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>YEAR</td>
                <td>Indicates the year when the outage event occurred</td>
            </tr>
            <tr>
                <td>MONTH</td>
                <td>Indicates the month when the outage event occurred</td>
            </tr>
            <tr>
                <td>U.S._STATE</td>
                <td>Represents all the states in the continental U.S</td>
            </tr>
            <tr>
                <td>ANOMALY.LEVEL</td>
                <td>This represents the oceanic El Niño/La Niña (ONI) index referring to the cold and warm episodes by season. It is estimated as a 3-month running mean of ERSST.v4 SST anomalies in the Niño 3.4 region (5°N to 5°S, 120–170°W)</td>
            </tr>
            <tr>
                <td>CLIMATE.CATEGORY</td>
                <td>This represents the climate episodes corresponding to the years. The categories—“Warm”, “Cold” or “Normal” episodes of the climate are based on a threshold of ± 0.5 °C for the Oceanic Niño Index (ONI)</td>
            </tr>
            <tr>
                <td>CAUSE.CATEGORY</td>
                <td>Categories of all the events causing the major power outages</td>
            </tr>
            <tr>
                <td>CAUSE.CATEGORY.DETAIL</td>
                <td>Detailed description of the event categories causing the major power outages</td>
            </tr>
            <tr>
                <td>OUTAGE.DURATION</td>
                <td>Duration of outage events (in minutes)</td>
            </tr>
            <tr>
                <td>DEMAND.LOSS.MW</td>
                <td>Amount of peak demand lost during an outage event (in Megawatt)</td>
            </tr>
            <tr>
                <td>CUSTOMERS.AFFECTED</td>
                <td>Number of customers affected by the power outage event</td>
            </tr>
            <tr>
                <td>RES.PRICE</td>
                <td>Electricity price for residencial use</td>
            </tr>
        </tbody>
    </table>

### Cleaning and EDA

#### Data Cleaning
We first throw away cloumns that we don't need for the analysis as mentioned above. We also drop 9 rows that critical infomation are missing.(We will not do any analysis on them since almost all variables are missing in those rows and it's hard for us to determine which type of missingness they belong to) We also assign some new variables(Season) to the dataframe for further analysis.

In [8]:
variable_names = [
    "YEAR",
    "MONTH",
    "U.S._STATE",
    "ANOMALY.LEVEL",
    "CLIMATE.CATEGORY",
    "CAUSE.CATEGORY",
    "CAUSE.CATEGORY.DETAIL",
    "OUTAGE.DURATION",
    "DEMAND.LOSS.MW",
    "CUSTOMERS.AFFECTED",
    "RES.PRICE"
]
df_cleaned = df[variable_names]
df_cleaned = df_cleaned[pd.notna(df_cleaned['MONTH'])]
def map_month_to_season(month):
    month = int(month)  # Convert float to int
    if month in [12, 1, 2]:
        return 'Winter'
    elif month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    else:
        return 'Fall'

# Apply the function to the MONTH column
df_cleaned['SEASON'] = df_cleaned['MONTH'].apply(map_month_to_season)
print(df_cleaned.head().to_markdown(index=False))

|   YEAR |   MONTH | U.S._STATE   |   ANOMALY.LEVEL | CLIMATE.CATEGORY   | CAUSE.CATEGORY     | CAUSE.CATEGORY.DETAIL   |   OUTAGE.DURATION |   DEMAND.LOSS.MW |   CUSTOMERS.AFFECTED |   RES.PRICE | SEASON   |
|-------:|--------:|:-------------|----------------:|:-------------------|:-------------------|:------------------------|------------------:|-----------------:|---------------------:|------------:|:---------|
|   2011 |       7 | Minnesota    |            -0.3 | normal             | severe weather     | nan                     |              3060 |              nan |                70000 |       11.6  | Summer   |
|   2014 |       5 | Minnesota    |            -0.1 | normal             | intentional attack | vandalism               |                 1 |              nan |                  nan |       12.12 | Spring   |
|   2010 |      10 | Minnesota    |            -1.5 | cold               | severe weather     | heavy wind              |              3000 |              nan |    

#### Univariate Analysis

We will use OUTAGE.DURATION, DEMAND.LOSS.MW, CUSTOMERS.AFFECTED to determine the seriousness of the power outage, so we will first look at their distribution. Since too many demand loss is missing, we will not consider this variable when determining the seriousness of the power outage.

In [170]:
nan_count = df_cleaned[['OUTAGE.DURATION', 'DEMAND.LOSS.MW', 'CUSTOMERS.AFFECTED']].isna().sum()
print("NaN Counts:\n", nan_count)


NaN Counts:
 OUTAGE.DURATION        49
DEMAND.LOSS.MW        702
CUSTOMERS.AFFECTED    441
dtype: int64


In [171]:
# Histogram for OUTAGE.DURATION
fig_duration = px.histogram(df_cleaned, x='OUTAGE.DURATION', title='OUTAGE.DURATION Distribution', nbins=40)
fig_duration.show()

# Histogram for CUSTOMERS.AFFECTED
fig_customers_affected = px.histogram(df_cleaned, x='CUSTOMERS.AFFECTED', title='CUSTOMERS.AFFECTED Distribution', nbins=40)
fig_customers_affected.show()

fig_duration.write_html('EDA-1.html',include_plotlyjs='cdn')
fig_customers_affected.write_html('EDA-2.html',include_plotlyjs='cdn')

We see some outliers that avoid us to see the details of the lower protion of the data, let's remove them for better representation.

In [172]:
def filter_outliers(column, lower_quantile=0.01, upper_quantile=0.95):
    lower_bound, upper_bound = column.quantile([lower_quantile, upper_quantile])
    return column.clip(lower=lower_bound, upper=upper_bound)

# Apply filtering to each column
filtered_duration = filter_outliers(df['OUTAGE.DURATION'])
filtered_customers_affected = filter_outliers(df['CUSTOMERS.AFFECTED'])

# Plotting histograms with filtered data
fig_duration_filtered = px.histogram(filtered_duration, title='OUTAGE.DURATION Distribution (Filtered)', nbins=40)
fig_customers_affected_filtered = px.histogram(filtered_customers_affected, title='CUSTOMERS.AFFECTED Distribution (Filtered)', nbins=40)

# Displaying the plots
fig_duration_filtered.show()
fig_customers_affected_filtered.show()

fig_duration_filtered.write_html('EDA-3.html',include_plotlyjs='cdn')
fig_customers_affected_filtered.write_html('EDA-4.html',include_plotlyjs='cdn')

We can now see how power outage duration and customer affected distribute clearly. We can choose a threshold for each data such that any outage that has worse impact than the threshold should be called a serious power outage. But before we do that, we will need to deal with the missingness of both variables above.

### Assessment of Missingness

#### Missingness of Duration

I believe the missingness of duration data is NMAR since it's more possible that the electricity companies would not publish a lot of details about the outage if the duration of it is too short. While there could be other reasons for the duration to be missing, such as some power companies never report the duration of the outage while others always report all the details, but since the main reason of missingness is still the value itself, we determine the missingness is NMAR (And there's hardly any extra data that could make it MAR), and we simply fill in 1 for all missingness(0 and NaN).

In [173]:
df_cleaned['OUTAGE.DURATION'] = df_cleaned['OUTAGE.DURATION'].apply(lambda x: 1 if (x==0 or pd.isna(x)) else x)
df_cleaned.head()

YEAR  MONTH U.S._STATE  ANOMALY.LEVEL CLIMATE.CATEGORY      CAUSE.CATEGORY  \
0  2011    7.0  Minnesota           -0.3           normal      severe weather   
1  2014    5.0  Minnesota           -0.1           normal  intentional attack   
2  2010   10.0  Minnesota           -1.5             cold      severe weather   
3  2012    6.0  Minnesota           -0.1           normal      severe weather   
4  2015    7.0  Minnesota            1.2             warm      severe weather   

  CAUSE.CATEGORY.DETAIL  OUTAGE.DURATION  DEMAND.LOSS.MW  CUSTOMERS.AFFECTED  \
0                   NaN           3060.0             NaN             70000.0   
1             vandalism              1.0             NaN                 NaN   
2            heavy wind           3000.0             NaN             70000.0   
3          thunderstorm           2550.0             NaN             68200.0   
4                   NaN           1740.0           250.0            250000.0   

   RES.PRICE  SEASON  
0      11.60  Summer  
1      12.12  Spring  
2      10.87    Fall  
3      11.79  Summer  
4      13.07  Summer

#### Missingness of Customer Affected
Missingness of customer affected could be dependent on the duration of power outage since the electricity company may not publish detailed information if the duration of power outage is too short. We will do a permutation test to prove it. Note here that Missingness of Customer Affected is also NMAR, since the missingness is also realted to its own value, but it also has a very strong correlation with duration and time(year) of the power outage, we will only focus on its correlation with duration here.

Null Hypothesis: The distribution of 'OUTAGE.DURATION' is independent of the missingness in the 'CUSTOMERS.AFFECTED' column. This means that the duration of outages does not influence whether the number of affected customers is reported.

Alterative Hypothesis: The 'OUTAGE.DURATION' tends to be lower (shorter) for cases where 'CUSTOMERS.AFFECTED' data is missing. This suggests that shorter outages are more likely to have missing data on the number of customers affected.

We will use the difference in mean duration as test statistic.

In [174]:
missing_affected = df_cleaned[['OUTAGE.DURATION', 'CUSTOMERS.AFFECTED']].copy()
missing_affected['missing_affected'] = pd.isna(missing_affected['CUSTOMERS.AFFECTED'])
observed = missing_affected.groupby('missing_affected')['OUTAGE.DURATION'].mean()
observed_stats = observed[False] - observed[True]
observed_stats

798.6881228506165

In [175]:
missing_affected_draw = missing_affected.copy()
missing_affected_draw['OUTAGE.DURATION'] = filter_outliers(missing_affected_draw['OUTAGE.DURATION'])
plot = px.histogram(missing_affected_draw, x='OUTAGE.DURATION', color='missing_affected', histnorm='probability', marginal='box',
             title="Outage Duration by Missingness of Customer Affected", barmode='overlay', opacity=0.6)
plot.show()
plot.write_html('MV-1.html',include_plotlyjs='cdn')

In [176]:
stats_test = []
for i in range(10000):
    test_permutation = np.random.permutation(missing_affected['missing_affected'])
    missing_affected['missing_affected'] = test_permutation
    test_observed = missing_affected.groupby('missing_affected')['OUTAGE.DURATION'].mean()
    test_stats = test_observed[False] - test_observed[True]
    stats_test.append(test_stats)
stats_test = np.array(stats_test)
print((stats_test > observed_stats).sum()/len(stats_test))

0.0028


We reject the null hypothesis with a alpha of 0.01, thus we can conclude that the missingness of CUSTOMERS.AFFECTED column is correlated with duration of the outage.

In [177]:
plot = px.histogram(stats_test, nbins = 50, histnorm='probability', title='Empirical Distribution of the mean difference')
plot.add_vline(observed_stats, line_color='red')
plot.add_annotation(text=f'<span style="color:red">Observed Difference = {round(observed_stats, 2)}</span>',
                   x=700, showarrow=False, y=0.05)
plot.show()
plot.write_html('MV-2.html',include_plotlyjs='cdn')

This suggests that the customer affected variable is much more likely to be missing if the duration of the power outage is short. We should keep this in mind when we determine whether a power outage is serious with these two variables.

Then, let's see whether the missingness has to do with a less relevant variable like RES.PRICE.

Null Hypothesis: The distribution of 'RES.PRICE' is independent of the missingness in the 'CUSTOMERS.AFFECTED' column. This means that the residential electricity price does not influence whether the number of affected customers is reported.

Alterative Hypothesis: The 'RES.PRICE' tends to be lower for cases where 'CUSTOMERS.AFFECTED' data is missing. This suggests companies that provides cheaper residential electricity are more likely to have missing data on the number of customers affected.

We will use the same statistics for this test.


In [178]:
missing_affected = df_cleaned[['RES.PRICE', 'CUSTOMERS.AFFECTED']].copy()
missing_affected['missing_affected'] = pd.isna(missing_affected['CUSTOMERS.AFFECTED'])
observed = missing_affected.groupby('missing_affected')['RES.PRICE'].mean()
observed_stats = observed[False] - observed[True]
observed_stats

-0.10206379649832442

In [179]:
plot = px.histogram(missing_affected, x='RES.PRICE', color='missing_affected', histnorm='probability', marginal='box',
             title="Residential Electricity Price by Missingness of Customer Affected", barmode='overlay', opacity=0.6)
plot.show()
plot.write_html('MV-3.html',include_plotlyjs='cdn')

In [180]:
stats_test = []
for i in range(10000):
    test_permutation = np.random.permutation(missing_affected['missing_affected'])
    missing_affected['missing_affected'] = test_permutation
    test_observed = missing_affected.groupby('missing_affected')['RES.PRICE'].mean()
    test_stats = test_observed[False] - test_observed[True]
    stats_test.append(test_stats)
stats_test = np.array(stats_test)
print((stats_test < observed_stats).sum()/len(stats_test))

0.284


We fail to reject the null hypothesis.

In [181]:
plot = px.histogram(stats_test, nbins = 50, histnorm='probability', title='Empirical Distribution of the mean difference')
plot.add_vline(observed_stats, line_color='red')
plot.add_annotation(text=f'<span style="color:red">Observed Difference = {round(observed_stats, 2)}</span>',
                   x=-0.17, showarrow=False, y=0.1)
plot.write_html('MV-4.html',include_plotlyjs='cdn')
plot.show()

Obviously the missingness of customer affected does not has strong correlation with the price of residential electricity price. This suggests we do not need to consider this variable when working with rows that has missing custormer affected.

### Back to EDA
Since we already know the missingness of customer affected is highly correlated with the duration of power outage, we can now consider how to define whether a outage is serious and thus we can do bivariate analysis to provide some insight for hypothesis testing later. We decide to call a outage serious if its duration is more than 2000 minutes and customer affected is more than 100,000. Since we know there's correlation between missingness of customer affected and duration, we will just ignore the condition on customer affected if it is missing.(We will say it is serious if the duration is more than 2000 minutes alone.)

In [182]:
df_cleaned['SERIOUSNESS'] = ((df_cleaned['OUTAGE.DURATION'] > 2000) & ((pd.isna(df_cleaned['CUSTOMERS.AFFECTED'])) | (df_cleaned['CUSTOMERS.AFFECTED'] > 100000)))
df_cleaned.head(10)

YEAR  MONTH U.S._STATE  ANOMALY.LEVEL CLIMATE.CATEGORY      CAUSE.CATEGORY  \
0  2011    7.0  Minnesota           -0.3           normal      severe weather   
1  2014    5.0  Minnesota           -0.1           normal  intentional attack   
2  2010   10.0  Minnesota           -1.5             cold      severe weather   
3  2012    6.0  Minnesota           -0.1           normal      severe weather   
4  2015    7.0  Minnesota            1.2             warm      severe weather   
5  2010   11.0  Minnesota           -1.4             cold      severe weather   
6  2010    7.0  Minnesota           -0.9             cold      severe weather   
7  2005    6.0  Minnesota            0.2           normal      severe weather   
8  2015    3.0  Minnesota            0.6             warm  intentional attack   
9  2013    6.0  Minnesota           -0.2           normal      severe weather   

  CAUSE.CATEGORY.DETAIL  OUTAGE.DURATION  DEMAND.LOSS.MW  CUSTOMERS.AFFECTED  \
0                   NaN           3060.0             NaN             70000.0   
1             vandalism              1.0             NaN                 NaN   
2            heavy wind           3000.0             NaN             70000.0   
3          thunderstorm           2550.0             NaN             68200.0   
4                   NaN           1740.0           250.0            250000.0   
5          winter storm           1860.0             NaN             60000.0   
6             tornadoes           2970.0             NaN             63000.0   
7          thunderstorm           3960.0            75.0            300000.0   
8              sabotage            155.0            20.0              5941.0   
9             hailstorm           3621.0             NaN            400000.0   

   RES.PRICE  SEASON  SERIOUSNESS  
0      11.60  Summer        False  
1      12.12  Spring        False  
2      10.87    Fall        False  
3      11.79  Summer        False  
4      13.07  Summer        False  
5      10.63    Fall        False  
6      11.41  Summer        False  
7       9.10  Summer         True  
8      11.53  Spring        False  
9      12.71  Summer         True

#### Bivariate Analysis
We now can see the proportion of serious power outage with different conditions set.

In [183]:
plot = px.bar(df_cleaned.groupby('CAUSE.CATEGORY')['SERIOUSNESS'].mean(), title = "Proportion of Serious Power Outage for Difference Outgae Causes")
plot.show()
plot.write_html('EDA-5.html',include_plotlyjs='cdn')


We can see severe weather conditions and fuel supply emergency tend to results in serious power outage. We will compare public appeal and severe weather in hypothesis testing later sicen they have more obvious difference.

In [184]:
plot = px.bar(df_cleaned.groupby('SEASON')['SERIOUSNESS'].mean(), title = "Proportion of Serious Power Outage for Difference Seasons")
plot.show()
plot.write_html('EDA-6.html',include_plotlyjs='cdn')

Obviously more serious power outage happens in Fall, probably related to some extreme weather conditions in that season. We will also do a hypothesis test comparing Fall and Spring in the later section.

In [185]:
plot = px.bar(df_cleaned.groupby('CLIMATE.CATEGORY')['SERIOUSNESS'].mean(), title = "Proportion of Serious Power Outage for Difference Climate Categories")
plot.show()
plot.write_html('EDA-7.html',include_plotlyjs='cdn')

Climate conditions of this type seems to be able to give us too much insight about serious power outages. We will skip this question in the hypothesis testing.

#### Interesting Aggregates

In [186]:
df_cleaned.pivot_table(index = "SEASON", columns= "CAUSE.CATEGORY", values = "SERIOUSNESS", aggfunc="mean")

CAUSE.CATEGORY  equipment failure  fuel supply emergency  intentional attack  \
SEASON                                                                         
Fall                     0.000000               0.000000            0.025316   
Spring                   0.058824               0.500000            0.045802   
Summer                   0.000000               0.333333            0.021277   
Winter                   0.000000               0.529412            0.017544   

CAUSE.CATEGORY  islanding  public appeal  severe weather  \
SEASON                                                     
Fall                  0.0       0.000000        0.436709   
Spring                0.0       0.000000        0.328000   
Summer                0.0       0.166667        0.282686   
Winter                0.0       0.181818        0.362694   

CAUSE.CATEGORY  system operability disruption  
SEASON                                         
Fall                                 0.000000  
Spring                               0.026316  
Summer                               0.133333  
Winter                               0.035714

Though some of the cause categories are obviously not relevant to season, there's still some important insights we can get from this pivot table. We can see fuel supply emergency usually cause much more trouble in Spring and Winter. We can also confirm that the reason why more serious power outage happens in Fall is because of severe weather.

### Hypothesis Testing

For the first hypothesis test, we will take a look at how season affects the seriousness of power outages.

Null hypothesis: Power outages happens in Spring and Fall are equally serious. (measured by the proportion of serious power outages out of all power outages.)
Alternative hypothesis: Power outages happens in Fall are more serious than in Spring. (measured by the proportion of serious power outages out of all power outages.)

I choose the difference in mean to be the test statistic, and significance level to be 0.01.

In [187]:
HT_test = df_cleaned[['SEASON', 'SERIOUSNESS']].copy()
HT_test = HT_test[(HT_test['SEASON'] == 'Fall') | (HT_test['SEASON'] == 'Spring')]
observed = HT_test.groupby('SEASON')['SERIOUSNESS'].mean()
observed_stats = observed['Fall'] - observed['Spring']
stats_test = []
for i in range(100000):
    test_permutation = np.random.permutation(HT_test['SERIOUSNESS'])
    HT_test['SERIOUSNESS'] = test_permutation
    test_observed = HT_test.groupby('SEASON')['SERIOUSNESS'].mean()
    test_stats = test_observed['Fall'] - test_observed['Spring']
    stats_test.append(test_stats)
stats_test = np.array(stats_test)
print((stats_test > observed_stats).sum()/len(stats_test))

0.00103


The resulting p-value is 0.001, thus we can reject the null hypothesis under the significance level of 0.01 and conclude that Power outages happens in Fall are generally more serious than in Spring.

For the second hypothesis test, we will take a look at how causes of outage affects the seriousness of power outages.

Null hypothesis: Power outages due to severe weather and public appeal are equally severe. (measured by the proportion of serious power outages out of all power outages.)
Alternative hypothesis: Power outages due to severe weather are more severe than outages due to public appeal. (measured by the proportion of serious power outages out of all power outages.)

I choose the difference in mean to be the test statistic, and significance level to be 0.01.

In [188]:
HT_test = df_cleaned[['CAUSE.CATEGORY', 'SERIOUSNESS']].copy()
HT_test = HT_test[(HT_test['CAUSE.CATEGORY'] == 'severe weather') | (HT_test['CAUSE.CATEGORY'] == 'public appeal')]
observed = HT_test.groupby('CAUSE.CATEGORY')['SERIOUSNESS'].mean()
observed_stats = observed['severe weather'] - observed['public appeal']
stats_test = []
for i in range(100000):
    test_permutation = np.random.permutation(HT_test['SERIOUSNESS'])
    HT_test['SERIOUSNESS'] = test_permutation
    test_observed = HT_test.groupby('CAUSE.CATEGORY')['SERIOUSNESS'].mean()
    test_stats = test_observed['severe weather'] - test_observed['public appeal']
    stats_test.append(test_stats)
stats_test = np.array(stats_test)
print((stats_test > observed_stats).sum()/len(stats_test))

0.00015


The resulting p-value is 0.00015, thus we can reject the null hypothesis under the significance level of 0.01 and conclude that Power outages due to severe weather are generally more severe than outages due to public appeal.